File: bots/minigames/cogs/flipcoin.py
"""
bots/minigames/cogs/flipcoin.py
─────────────────────────────────
/flipcoin – Coin flip game with Heads/Tails buttons and points integration.
"""

from __future__ import annotations

import logging
import random

import discord
from discord import app_commands
from discord.ext import commands
from discord.ui import Button, View

from core.points_manager import add_points, get_pts, notify_rewards, points_footer
from core.utils import make_embed

log = logging.getLogger(__name__)


# ─── View ─────────────────────────────────────────────────────────────────────

class FlipCoinView(View):
    def __init__(self, player: discord.User) -> None:
        super().__init__(timeout=60)
        self.player = player

    @discord.ui.button(label="🪙 Heads", style=discord.ButtonStyle.primary,  custom_id="flip_heads")
    async def heads(self, interaction: discord.Interaction, button: Button) -> None:
        await self._flip(interaction, "heads")

    @discord.ui.button(label="🔙 Tails", style=discord.ButtonStyle.secondary, custom_id="flip_tails")
    async def tails(self, interaction: discord.Interaction, button: Button) -> None:
        await self._flip(interaction, "tails")

    async def _flip(self, interaction: discord.Interaction, choice: str) -> None:
        if interaction.user.id != self.player.id:
            await interaction.response.send_message("❌ This is not your game!", ephemeral=True)
            return

        result = random.choice(["heads", "tails"])
        won    = result == choice

        # Punkte vergeben
        outcome   = "win" if won else "lose"
        pts_delta = get_pts("flipcoin", outcome)
        old, new  = await add_points(interaction.user.id, pts_delta)

        for item in self.children:
            item.disabled = True
        self.stop()

        embed = make_embed(
            title="🪙 Flip a Coin",
            description=(
                f"You chose: **{choice.capitalize()}**\n"
                f"Result:    **{result.capitalize()}**\n\n"
                + ("You won! 🎉" if won else "You lost. 😔")
            ),
            footer_text=f"Flip a Coin  •  {points_footer(pts_delta, new)}",
        )
        await interaction.response.edit_message(embed=embed, view=self)
        await notify_rewards(interaction, old, new)

    async def on_timeout(self) -> None:
        for item in self.children:
            item.disabled = True


# ─── Cog ──────────────────────────────────────────────────────────────────────

class FlipCoinCog(commands.Cog, name="FlipCoin"):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot

    @app_commands.command(name="flipcoin", description="Flip a coin – Heads or Tails?")
    async def flipcoin(self, interaction: discord.Interaction) -> None:
        embed = make_embed(
            title="🪙 Flip a Coin",
            description=f"{interaction.user.mention}, choose your side!",
        )
        await interaction.response.send_message(embed=embed, view=FlipCoinView(interaction.user))


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(FlipCoinCog(bot))

File: bots/minigames/commands/flipcoin.js
const { SlashCommandBuilder, ButtonBuilder, ButtonStyle, ActionRowBuilder } = require('discord.js');
const { makeEmbed } = require('../../../core/utils');
const { applyMeta } = require('../../../core/commandKit');
const { gameFooter } = require('../../../core/gameKit');
const { addPoints, getPts, notifyRewards } = require('../../../core/pointsManager');
const { t } = require('../../../core/i18n');

module.exports = {
  key: 'flipcoin',
  game: 'flipcoin',
  data: applyMeta(new SlashCommandBuilder(), 'flipcoin'),

  async execute(interaction) {
    const embed = makeEmbed({
      title:       t('games.flipcoin.title'),
      description: t('games.flipcoin.choose', { user: String(interaction.user) }),
    });

    const row = new ActionRowBuilder().addComponents(
      new ButtonBuilder().setCustomId('flip_heads').setLabel(t('games.flipcoin.headsButton')).setStyle(ButtonStyle.Primary),
      new ButtonBuilder().setCustomId('flip_tails').setLabel(t('games.flipcoin.tailsButton')).setStyle(ButtonStyle.Secondary),
    );

    await interaction.reply({ embeds: [embed], components: [row] });
    const reply = await interaction.fetchReply();

    const collector = reply.createMessageComponentCollector({
      filter: i => i.user.id === interaction.user.id,
      time: 60_000, max: 1,
    });

    collector.on('collect', async i => {
      const choice = i.customId === 'flip_heads' ? 'heads' : 'tails';
      const result = Math.random() < 0.5 ? 'heads' : 'tails';
      const won = choice === result;

      const delta = getPts('flipcoin', won ? 'win' : 'lose');
      const { old: oldPts, new: newPts } = await addPoints(interaction.user.id, delta);

      for (const btn of row.components) btn.setDisabled(true);

      const resultEmbed = makeEmbed({
        title: t('games.flipcoin.title'),
        description: t('games.flipcoin.result', {
          choice:  t(`games.flipcoin.${choice}`),
          result:  t(`games.flipcoin.${result}`),
          outcome: won ? t('games.flipcoin.won') : t('games.flipcoin.lost'),
        }),
        footerText: gameFooter('flipcoin', { delta, total: newPts }),
      });

      await i.update({ embeds: [resultEmbed], components: [row] });
      await notifyRewards(i, oldPts, newPts);
    });

    collector.on('end', (collected) => {
      if (!collected.size) {
        for (const btn of row.components) btn.setDisabled(true);
        interaction.editReply({ components: [row] }).catch(() => {});
      }
    });
  },
};
